Make TestWebKitAPI::WebTransportServer more robust

achristensen07 opened the pull request:
#### 72de008ffc2a35922d34a064338028ea694ff7b6
<pre>
Make TestWebKitAPI::WebTransportServer more robust
<a href="https://bugs.webkit.org/show_bug.cgi?id=279107">https://bugs.webkit.org/show_bug.cgi?id=279107</a>
<a href="https://rdar.apple.com/135249015">rdar://135249015</a>

Reviewed by Matthew Finkel.

WebTransportServer was barely enough to do one echo.
This makes it more robust.  It now receives a connection
handler in its constructor and shares the TestWebKitAPI::Connection
code with HTTPServer.  This will allow us to write more interesting
tests.

* Tools/TestWebKitAPI/CoroutineUtilities.h: Copied from Tools/TestWebKitAPI/WebTransportServer.h.
(TestWebKitAPI::CoroutineHandle::CoroutineHandle):
(TestWebKitAPI::CoroutineHandle::~CoroutineHandle):
(TestWebKitAPI::Task::promise_type::get_return_object):
(TestWebKitAPI::Task::promise_type::initial_suspend):
(TestWebKitAPI::Task::promise_type::unhandled_exception):
(TestWebKitAPI::Task::promise_type::return_void):
* Tools/TestWebKitAPI/NetworkConnection.h: Added.
(TestWebKitAPI::Connection::receiveHTTPRequest):
(TestWebKitAPI::Connection::webSocketHandshake):
(TestWebKitAPI::Connection::terminate):
(TestWebKitAPI::Connection::Connection):
(TestWebKitAPI::ReceiveHTTPRequestOperation::ReceiveHTTPRequestOperation):
(TestWebKitAPI::ReceiveHTTPRequestOperation::await_ready):
(TestWebKitAPI::ReceiveHTTPRequestOperation::await_resume):
(TestWebKitAPI::ReceiveBytesOperation::ReceiveBytesOperation):
(TestWebKitAPI::ReceiveBytesOperation::await_ready):
(TestWebKitAPI::ReceiveBytesOperation::await_resume):
(TestWebKitAPI::SendOperation::SendOperation):
(TestWebKitAPI::SendOperation::await_ready):
(TestWebKitAPI::SendOperation::await_resume):
* Tools/TestWebKitAPI/NetworkConnection.mm: Added.
(TestWebKitAPI::dataFromVector):
(TestWebKitAPI::vectorFromData):
(TestWebKitAPI::dataFromString):
(TestWebKitAPI::Connection::receiveBytes const):
(TestWebKitAPI::Connection::receiveHTTPRequest const):
(TestWebKitAPI::Connection::awaitableReceiveHTTPRequest const):
(TestWebKitAPI::ReceiveHTTPRequestOperation::await_suspend):
(TestWebKitAPI::Connection::awaitableReceiveBytes const):
(TestWebKitAPI::ReceiveBytesOperation::await_suspend):
(TestWebKitAPI::SendOperation::await_suspend):
(TestWebKitAPI::Connection::awaitableSend):
(TestWebKitAPI::Connection::send const):
(TestWebKitAPI::Connection::sendAndReportError const):
(TestWebKitAPI::Connection::webSocketHandshake):
(TestWebKitAPI::Connection::terminate):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, Basic)):
* Tools/TestWebKitAPI/WebTransportServer.h:
* Tools/TestWebKitAPI/WebTransportServer.mm:
(TestWebKitAPI::WebTransportServer::Data::create):
(TestWebKitAPI::WebTransportServer::Data::Data):
(TestWebKitAPI::WebTransportServer::WebTransportServer):
(TestWebKitAPI::WebTransportServer::port const):
* Tools/TestWebKitAPI/cocoa/HTTPServer.h:
(TestWebKitAPI::CoroutineHandle::CoroutineHandle): Deleted.
(TestWebKitAPI::CoroutineHandle::~CoroutineHandle): Deleted.
(TestWebKitAPI::Task::promise_type::get_return_object): Deleted.
(TestWebKitAPI::Task::promise_type::initial_suspend): Deleted.
(TestWebKitAPI::Task::promise_type::unhandled_exception): Deleted.
(TestWebKitAPI::Task::promise_type::return_void): Deleted.
(TestWebKitAPI::Connection::receiveHTTPRequest): Deleted.
(TestWebKitAPI::Connection::webSocketHandshake): Deleted.
(TestWebKitAPI::Connection::terminate): Deleted.
(TestWebKitAPI::Connection::Connection): Deleted.
(TestWebKitAPI::ReceiveOperation::ReceiveOperation): Deleted.
(TestWebKitAPI::ReceiveOperation::await_ready): Deleted.
(TestWebKitAPI::ReceiveOperation::await_resume): Deleted.
(TestWebKitAPI::SendOperation::SendOperation): Deleted.
(TestWebKitAPI::SendOperation::await_ready): Deleted.
(TestWebKitAPI::SendOperation::await_resume): Deleted.
* Tools/TestWebKitAPI/cocoa/HTTPServer.mm:
(TestWebKitAPI::dataFromString): Deleted.
(TestWebKitAPI::dataFromVector): Deleted.
(TestWebKitAPI::vectorFromData): Deleted.
(TestWebKitAPI::Connection::receiveBytes const): Deleted.
(TestWebKitAPI::Connection::receiveHTTPRequest const): Deleted.
(TestWebKitAPI::Connection::awaitableReceiveHTTPRequest const): Deleted.
(TestWebKitAPI::ReceiveOperation::await_suspend): Deleted.
(TestWebKitAPI::SendOperation::await_suspend): Deleted.
(TestWebKitAPI::Connection::awaitableSend): Deleted.
(TestWebKitAPI::Connection::send const): Deleted.
(TestWebKitAPI::Connection::sendAndReportError const): Deleted.
(TestWebKitAPI::Connection::webSocketHandshake): Deleted.
(TestWebKitAPI::Connection::terminate): Deleted.

Canonical link: <a href="https://commits.webkit.org/283177@main">https://commits.webkit.org/283177@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/67308581d106f53d413f555efd55b034417df838

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65480 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44851 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18098 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69506 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16089 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67598 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52652 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16369 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/11156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68547 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/41441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/56684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33207 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/38116 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/14059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14965 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/14400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71211 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9434 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/13848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9466 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/56745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/60172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7797 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/1456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9927 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/40661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/41737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/42920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41481 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on Tools/TestWebKitAPI/NetworkConnection.h:
```diff
@@ -0,0 +1,103 @@
+class ReceiveBytesOperation;
+class SendOperation;
+
+class Connection {
```
Might want to rename this in a follow-up, too.

achristensen07 replied:
Maybe

sysrqb commented on Tools/TestWebKitAPI/WebTransportServer.mm:
```diff
@@ -29,11 +29,27 @@
+WebTransportServer::WebTransportServer(Function<Task(Connection)>&& connectionHandler)
+    : m_data(Data::create(WTFMove(connectionHandler)))
 {
+    Ref data { m_data };
```
`protectedData`?

achristensen07 replied:
This is only needed for the block captures.  I think I'll use lambdas instead and capture data more explicitly.

sysrqb reviewed (approved):
Seems okay. I'll trust you that the Connection class was basically copy-and-pasted.

achristensen07 committed:
Make TestWebKitAPI::WebTransportServer more robust
https://bugs.webkit.org/show_bug.cgi?id=279107
rdar://135249015

Reviewed by Matthew Finkel.

WebTransportServer was barely enough to do one echo.
This makes it more robust.  It now receives a connection
handler in its constructor and shares the TestWebKitAPI::Connection
code with HTTPServer.  This will allow us to write more interesting
tests.

* Tools/TestWebKitAPI/CoroutineUtilities.h: Copied from Tools/TestWebKitAPI/WebTransportServer.h.
(TestWebKitAPI::CoroutineHandle::CoroutineHandle):
(TestWebKitAPI::CoroutineHandle::~CoroutineHandle):
(TestWebKitAPI::Task::promise_type::get_return_object):
(TestWebKitAPI::Task::promise_type::initial_suspend):
(TestWebKitAPI::Task::promise_type::unhandled_exception):
(TestWebKitAPI::Task::promise_type::return_void):
* Tools/TestWebKitAPI/NetworkConnection.h: Added.
(TestWebKitAPI::Connection::receiveHTTPRequest):
(TestWebKitAPI::Connection::webSocketHandshake):
(TestWebKitAPI::Connection::terminate):
(TestWebKitAPI::Connection::Connection):
(TestWebKitAPI::ReceiveHTTPRequestOperation::ReceiveHTTPRequestOperation):
(TestWebKitAPI::ReceiveHTTPRequestOperation::await_ready):
(TestWebKitAPI::ReceiveHTTPRequestOperation::await_resume):
(TestWebKitAPI::ReceiveBytesOperation::ReceiveBytesOperation):
(TestWebKitAPI::ReceiveBytesOperation::await_ready):
(TestWebKitAPI::ReceiveBytesOperation::await_resume):
(TestWebKitAPI::SendOperation::SendOperation):
(TestWebKitAPI::SendOperation::await_ready):
(TestWebKitAPI::SendOperation::await_resume):
* Tools/TestWebKitAPI/NetworkConnection.mm: Added.
(TestWebKitAPI::dataFromVector):
(TestWebKitAPI::vectorFromData):
(TestWebKitAPI::dataFromString):
(TestWebKitAPI::Connection::receiveBytes const):
(TestWebKitAPI::Connection::receiveHTTPRequest const):
(TestWebKitAPI::Connection::awaitableReceiveHTTPRequest const):
(TestWebKitAPI::ReceiveHTTPRequestOperation::await_suspend):
(TestWebKitAPI::Connection::awaitableReceiveBytes const):
(TestWebKitAPI::ReceiveBytesOperation::await_suspend):
(TestWebKitAPI::SendOperation::await_suspend):
(TestWebKitAPI::Connection::awaitableSend):
(TestWebKitAPI::Connection::send const):
(TestWebKitAPI::Connection::sendAndReportError const):
(TestWebKitAPI::Connection::webSocketHandshake):
(TestWebKitAPI::Connection::terminate):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, Basic)):
* Tools/TestWebKitAPI/WebTransportServer.h:
* Tools/TestWebKitAPI/WebTransportServer.mm:
(TestWebKitAPI::WebTransportServer::Data::create):
(TestWebKitAPI::WebTransportServer::Data::Data):
(TestWebKitAPI::WebTransportServer::WebTransportServer):
(TestWebKitAPI::WebTransportServer::port const):
* Tools/TestWebKitAPI/cocoa/HTTPServer.h:
(TestWebKitAPI::CoroutineHandle::CoroutineHandle): Deleted.
(TestWebKitAPI::CoroutineHandle::~CoroutineHandle): Deleted.
(TestWebKitAPI::Task::promise_type::get_return_object): Deleted.
(TestWebKitAPI::Task::promise_type::initial_suspend): Deleted.
(TestWebKitAPI::Task::promise_type::unhandled_exception): Deleted.
(TestWebKitAPI::Task::promise_type::return_void): Deleted.
(TestWebKitAPI::Connection::receiveHTTPRequest): Deleted.
(TestWebKitAPI::Connection::webSocketHandshake): Deleted.
(TestWebKitAPI::Connection::terminate): Deleted.
(TestWebKitAPI::Connection::Connection): Deleted.
(TestWebKitAPI::ReceiveOperation::ReceiveOperation): Deleted.
(TestWebKitAPI::ReceiveOperation::await_ready): Deleted.
(TestWebKitAPI::ReceiveOperation::await_resume): Deleted.
(TestWebKitAPI::SendOperation::SendOperation): Deleted.
(TestWebKitAPI::SendOperation::await_ready): Deleted.
(TestWebKitAPI::SendOperation::await_resume): Deleted.
* Tools/TestWebKitAPI/cocoa/HTTPServer.mm:
(TestWebKitAPI::dataFromString): Deleted.
(TestWebKitAPI::dataFromVector): Deleted.
(TestWebKitAPI::vectorFromData): Deleted.
(TestWebKitAPI::Connection::receiveBytes const): Deleted.
(TestWebKitAPI::Connection::receiveHTTPRequest const): Deleted.
(TestWebKitAPI::Connection::awaitableReceiveHTTPRequest const): Deleted.
(TestWebKitAPI::ReceiveOperation::await_suspend): Deleted.
(TestWebKitAPI::SendOperation::await_suspend): Deleted.
(TestWebKitAPI::Connection::awaitableSend): Deleted.
(TestWebKitAPI::Connection::send const): Deleted.
(TestWebKitAPI::Connection::sendAndReportError const): Deleted.
(TestWebKitAPI::Connection::webSocketHandshake): Deleted.
(TestWebKitAPI::Connection::terminate): Deleted.

Canonical link: https://commits.webkit.org/283177@main
